# Homework 6: Due Wednesday, March 12 at 11:59 PM

---

## <a name="HW1Inst">Instructions:</a>
---

1. <font color="tomato">Either click the COPY TO DRIVE button or use the menu File/Save/Save a Copy to save your own version of the notebook in your own folder in Google Drive.</font>

  - If you do not have a Google account, you will need to create a Google account in order to create your own individual copy of this notebook.
  - By default, the notebook will be saved in a folder named **Colab Notebooks** in your drive.
  - Feel free to name and store the notebook wherever you like!

2. <font color="mediumseagreen">After you have copied the notebook into your Drive, you can begin editing and saving your work.</font>

**Submit your work electronically as two separate files:**

- A Google Colab (or Jupyter) Notebook with file extension `.ipynb`.
- A `covid-hw6.csv` file with final tidy version of the data.

Upload your file into the Homework 6 assignment in Canvas before <font color="dodgerblue">**11:59PM on Wednesday, March 12.**</font>

**You must provide supporting work for your all of your answers.**

  - <font color="dodgerblue">That work will include using R code cells.</font>
    - **Feel free to insert as much R code as you like.**
    - You must interpret or explain how the output of a code cell is being utilized to solve the problem.
    - **Code with incomplete or missing explanations of why the output is useful may not receive full credit.**
  - <font color="dodgerblue">Your work will also involve writing explanations and justifications in text cells.</font>
    - **Be sure to type/write out the steps of your work/explanation so I can see how you arrived at your answer.**


## Important Academic Policies:
---

- **Be your own work.** Though you may collaborate with others, everyone is responsible for writing up the work in their own way using their own methods. Plagiarism of any form is not tolerated.
- **Be complete.** You must provide all work and/or explanations needed to find the solution. Answers with insufficient or incomplete supporting work may lose credit.
- **Adhere to the Code of Academic Honesty.**
- **Be clearly written (and legible if written).** Your solution to a problem must be clear, written in complete sentences. You may lose credit for work that is unclear or hard to follow.

## <font color="dodgerblue">Let me know if you need help, and **GOOD LUCK!!!**</font>

----

# <a name="intro">Introduction</a>

---


In this assignment, you will be combining three different data sets into a single clean data set. You will be combining two data sets related to census demographics with a COVID-related data set and a voting-related data set.

The file `acs_pop_inc.csv` includes estimates of population and income for each of the 50 U.S. states (and the District of Columbia) based on the 2017 American Community Survey. The included variables for each observation are:

- `GEOID`: a unique number identifier associated with each state.
- `NAME`: the name of the state
- `variable`: a character string indicating what each value represents
0 `estimate`: the estimated value

The file `census_2010_education.csv` includes estimates of education levels for each of the 50 U.S. states (and the District of Columbia) based on the 2010 U.S. Census. The included variables are:

- `state`: the name of the state
- `percent_hs_estimate`: the percentage of the population 25 years or older with a high school degree
- `percent_bs_estimate`: the percentage of the population 25 years or older with a bachelor's degree

The file `all-states-covid.csv` includes data related to COVID-19 deaths, positive cases, etc. for each of the 50 U.S. states (and territories) as of March 7, 2021. The variables we will consider are:

- `date`: the date the observation was observed
- `state`: the name of the state or territory
- `death`: the total number of confirmed or probable COVID-19 deaths since the COVID-19 pandemic started
- `positive`: the total number of confirmed or probable positive COVID-19 cases since the COVID-19 pandemic started

<font color="dodgerblue">**You will need to load the `tidyverse` core packages each time you start a new Colab session to work on the problems in this assignment:**</font>





In [27]:
library(tidyverse)

# <a name="Q1">Question 1</a>

---

Run first the code cell below to use the `read_csv()` function from the `readr` package in the `tidyverse` to import the data provided in file `acs_pop_inc.csv` and assign it to the tibble `pop_inc`.

Then run an additional command(s) to see the structure of how data is initially stored in the data frame `pop_inc`, **without printing the entire data frame to the screen**. Points will be deducted for displaying a large amount of information to the screen.

Based on the output you generate, in the text cell below, explain why `pop_inc` is not a tidy data set. Do not simply quote the rules it violates. For this particular data set, clearly explain which variables/observations are not tidy. Be sure to write in complete sentences.


<br>

<font color="dodgerblue">*Note: The `read_csv` function is an enhanced version of `base` R's `read.csv`. The `read_csv` function is faster and imports the data frame as a tibble which has more functionality in the `tidyverse`!*</font>


In [28]:
# run code cell to load data to data frame named pop_inc
# no output will be generated after running this code cell
pop_inc <- read_csv(
  file = "https://raw.githubusercontent.com/CU-Denver-MathStats-OER/Data-Wrangling-and-Visualization/main/Data/acs_pop_inc.csv",
  show_col_types = FALSE
)

In [29]:
# use this code cell to generate output to answer the question
# do not print the entire data frame to the screen!
# points will be deducted for displaying too much info to the screen

# I decided to use print(df, num_rows) to print out the first 5 rows of the data frame to be able to see what variables are not tidy
print(pop_inc, n = 5)

# A tibble: 104 × 4
  GEOID NAME    variable      estimate
  <dbl> <chr>   <chr>            <dbl>
1     1 Alabama population     4876250
2     1 Alabama median_income    26596
3     2 Alaska  population      737068
4     2 Alaska  median_income    35957
5     4 Arizona population     7050299
# ℹ 99 more rows


## <a name="Sol1">Solution to Question 1</a>

---
There are 2 values in the "variable" column, these values should be in a separate column which makes this variable/column not tidy.

And since there are 2 values in the variable columns (population and median income), it makes that each observation has 2 rows. For example, the "Alabama" observations has 2 rows corresponding to Alabama, which makes the rows in this observations not tidy.

For this data frame to become tidy, there should be a population column and median_income column. Because when theres these 2 columns, the values would corresponds to only one observation. For example, after making those 2 columns, there would be only 1 Alabama and not two Alabama in the same table.

(I do know that this is a tibble, but I'm using the word "data frame" to explain because I feel like it is easier to understand when using the word data frame instead of tibble)





<br>  
<br>  


# <a name="Q2">Question 2</a>

---

Transform `pop_inc` into a tidy data frame. Assign the resulting tidy data frame the name `pop_inc_tidy`. Print the first 5 rows of `pop_inc_tidy` to the screen.


In [30]:
# Use code cell to answer question

# Clean the data frame by creating two new columns using pivot_wider(), and store it in pop_inc_tidy
pop_inc_tidy <- pivot_wider(pop_inc, names_from = variable,
                            values_from = estimate)

# Print out the first 5 rows of the tidied data frame
print(pop_inc_tidy, n = 5)

# A tibble: 52 × 4
  GEOID NAME       population median_income
  <dbl> <chr>           <dbl>         <dbl>
1     1 Alabama       4876250         26596
2     2 Alaska         737068         35957
3     4 Arizona       7050299         30552
4     5 Arkansas      2999370         26137
5     6 California   39283497         32232
# ℹ 47 more rows


# <a name="Q3">Question 3</a>

---

The `GEOID` column of `pop_inc_tidy` is not useful because the `NAME` column unique identifies each observation. Use the `select()` function to select all columns of `pop_inc_tidy` except `GEOID`.  Assign the resulting object back to `pop_inc_tidy`. Print the first 5 rows of `pop_inc_tidy`.


In [31]:
# Use code cell to answer question

# Use piping (optional) to tell the select() function to use pop_inc_tidy's data frame, and !GEOID means select every columns that is not GEOID, and store it in pop_inc_tidy
pop_inc_tidy <- pop_inc_tidy |> select(!GEOID)

# Print out the first 5 rows to check the results of the new data frame
print(pop_inc_tidy, n = 5)

# A tibble: 52 × 3
  NAME       population median_income
  <chr>           <dbl>         <dbl>
1 Alabama       4876250         26596
2 Alaska         737068         35957
3 Arizona       7050299         30552
4 Arkansas      2999370         26137
5 California   39283497         32232
# ℹ 47 more rows


# <a name="Q4">Question 4</a>

---

In each of the four code cells below, use functions from the `dplyr` that we have discussed so far this semester to display the rows in `pop_inc_tidy` corresponding to the states with:

-  The three greatest median incomes.
-  The three smallest median incomes.
-  The four greatest populations.
-  The four smallest populations.

**Do not assign the output of any of the commands to any object.** There is no unique way to accomplish this task (multiple methods are possible), but no credit will be earned for functions used that are not from the `dplyr` package that we have not discussed in our course materials this semester.

In [32]:
# print the rows with the 3 greatest median incomes

# Used slice_max() to find the greatest median incomes, and n just means how many rows I want
pop_inc_tidy |> slice_max(median_income, n = 3)

NAME                 population median_income
1 District of Columbia  692683    50134        
2 Maryland             6018848    41053        
3 New Jersey           8878503    39078

In [33]:
# print the rows with the 3 smallest median incomes

# Used slice_min() to find the smallest median incomes
pop_inc_tidy |> slice_min(median_income, n = 3)

NAME          population median_income
1 Puerto Rico   3318447    13641        
2 Mississippi   2984418    24924        
3 West Virginia 1817305    25669

In [34]:
# print the rows with the 4 greatest populations

# Used slice_max() to find the greatest population
pop_inc_tidy |> slice_max(population, n = 4)

NAME       population median_income
1 California 39283497   32232        
2 Texas      28260856   30993        
3 Florida    20901636   28291        
4 New York   19572319   34378

In [35]:
# print the rows with the 4 smallest populations

# Used slice_min() to find the smallest population
pop_inc_tidy |> slice_min(population, n = 4)

NAME                 population median_income
1 Wyoming              581024     32660        
2 Vermont              624313     32055        
3 District of Columbia 692683     50134        
4 Alaska               737068     35957

# <a name="Q5">Question 5</a>

---

Use the `read_csv` function from the `readr` package to import the data provided in `census_2010_education.csv`. Assign the data frame the name `educ`. The first 4 rows will be displayed on screen.

Then in the blank code cell, use the `arrange()` function to sort the rows of `educ` in descending order by `percent_bs_estimate`.

- Assign the resulting object to the same name `educ`.
- Print the first 5 rows of the new, sorted `educ`.





In [36]:
# import education data from external csv file
educ <- read_csv(
  file = "https://raw.githubusercontent.com/CU-Denver-MathStats-OER/Data-Wrangling-and-Visualization/main/Data/census_2010_education.csv",
  show_col_types = FALSE
)
print(educ, n = 4)

# A tibble: 51 × 3
  state    percent_hs_estimate percent_bs_estimate
  <chr>                  <dbl>               <dbl>
1 Alabama                 82.1                21.9
2 Alaska                  91                  27.9
3 Arizona                 85.6                25.9
4 Arkansas                82.9                19.5
# ℹ 47 more rows


In [37]:
# sort the rows of educ in descending order by percent_bs_estimate
# assign the result back to educ
# print the first 5 rows to the screen

# Used piping to tell the arrange() function to use educ's data frame, and use arrange() to sort the rows, and desc() to sort in descending order, and store it in educ
educ <- educ |> arrange(desc(percent_bs_estimate))

# Print out the first 5 rows of the result
print(educ, n = 5)

# A tibble: 51 × 3
  state                percent_hs_estimate percent_bs_estimate
  <chr>                              <dbl>               <dbl>
1 District of Columbia                87.4                50.1
2 Massachusetts                       89.1                39  
3 Colorado                            89.7                36.4
4 Maryland                            88.1                36.1
5 Connecticut                         88.6                35.5
# ℹ 46 more rows


# <a name="Q6">Question 6</a>

---

Use the `base` R function `merge()` to join the `pop_inc_tidy` (created in [Question 3](#Q3)) and `educ` (created in [Question 5](#Q5)) data frames by state name.

- Assign the new, merged data frame the name `state_demo`.
- Print the first 6 rows of `state_demo` using the `head()` function (because `state_demo` is a `data.frame` and not a tibble).

<br>

*Note: You cannot do this `merge()` automatically because the name of the column containing the state names differs for the two data frames. Run `?merge` for more information about the `merge()` function to determine how to resolve that issue. Or you can get creative and think of your own way of resolving the issue.*




In [38]:
# Ran this code to see merge()'s function arguments
?merge

In [39]:
# merge pop_inc_tidy and educ data frames
# store result to state_demo
# display first 6 rows of state_demo on screen

# Use merge() to merge the columns, and to merge specific columns, I used by.x = 1 which means pick the column 1 from x where x is pop_inc_tidy, and by.y does the same thing but for educ,
# and store it in state_demo
state_demo <- merge(pop_inc_tidy, educ, by.x = 1, by.y = 1)

# Use head() to print out the first 6 rows of the data frame to check the result
head(state_demo)

NAME       population median_income percent_hs_estimate percent_bs_estimate
1 Alabama     4876250   26596         82.1                21.9               
2 Alaska       737068   35957         91.0                27.9               
3 Arizona     7050299   30552         85.6                25.9               
4 Arkansas    2999370   26137         82.9                19.5               
5 California 39283497   32232         80.7                30.1               
6 Colorado    5610349   36644         89.7                36.4

# <a name="Q7">Question 7</a>

---

Use the `rename()` function to rename the `NAME` column of `state_demo` to `state`.

- Assign the updated data frame back to the same name `state_demo`.
- Use the `head` function to print the first 6 rows of `state_demo` to check your work.

<br>




In [40]:
# use code cell to answer question

# Used piping to tell the rename() function that I am using state_demo's data frame, and use rename(new_name = old_name) to change the column's name, and store it in state_demo
state_demo <- state_demo |> rename(state = NAME)

# Use head() to print out the first 6 rows of the data frame to check the result
head(state_demo)

state      population median_income percent_hs_estimate percent_bs_estimate
1 Alabama     4876250   26596         82.1                21.9               
2 Alaska       737068   35957         91.0                27.9               
3 Arizona     7050299   30552         85.6                25.9               
4 Arkansas    2999370   26137         82.9                19.5               
5 California 39283497   32232         80.7                30.1               
6 Colorado    5610349   36644         89.7                36.4

# <a name="Q8">Question 8</a>

---

Use the `read_csv` function from the `readr` package to import the data provided in `all-states-covid.csv`. Assign the data frame the name `covid`. A `glimpse()` of the data frame will be printed to the screen so you can see the full list of all initial variables in the data.

Then in the blank code cell, use the `select()` function to select the `date`, `state`, `death`, and `positive` columns of the `covid` data.


- Assign the updated data frame back to the same name `covid`.
- Print the first 5 rows of `covid` to the screen.





In [41]:
# import covid data from external csv file
# ignore warning message generated in output
covid <- read_csv(
  file = "https://raw.githubusercontent.com/CU-Denver-MathStats-OER/Data-Wrangling-and-Visualization/main/Data/all-states-covid.csv",
  show_col_types = FALSE
)
glimpse(covid)  # a glimpse of the variables in the data

Warning message:
“One or more parsing issues, call `problems()` on your data frame for details,
e.g.:
  dat <- vroom(...)
  problems(dat)”


Rows: 20,780
Columns: 41
$ date                             <chr> "3/7/2021", "3/7/2021", "3/7/2021", "…
$ state                            <chr> "AK", "AL", "AR", "AS", "AZ", "CA", "…
$ death                            <dbl> 305, 10148, 5319, 0, 16328, 54124, 59…
$ deathConfirmed                   <dbl> NA, 7963, 4308, NA, 14403, NA, 5251, …
$ deathIncrease                    <dbl> 0, -1, 22, 0, 5, 258, 3, 0, 0, 9, 66,…
$ deathProbable                    <dbl> NA, 2185, 1011, NA, 1925, NA, 735, 13…
$ hospitalized                     <dbl> 1293, 45976, 14926, NA, 57907, NA, 23…
$ hospitalizedCumulative           <dbl> 1293, 45976, 14926, NA, 57907, NA, 23…
$ hospitalizedCurrently            <dbl> 33, 494, 335, NA, 963, 4291, 326, 428…
$ hospitalizedIncrease             <dbl> 0, 0, 11, 0, 44, 0, 18, 0, 0, 0, 92, …
$ inIcuCumulative                  <dbl> NA, 2676, NA, NA, NA, NA, NA, NA, NA,…
$ inIcuCurrently                   <dbl> NA, NA, 141, NA, 273, 1159, NA, NA, 3…
$ negative     

In [42]:
# select date, state, death, and positive columns
# assign the result back to covid
# print the first 5 rows to the screen

# Use select() to select the columns that I want (From now I will omit explaining on using piping), and store it in covid
covid <- covid |> select(date, state, death, positive)

# Print out the first 5 rows of the new data frame to check the result
print(covid, n = 5)

# A tibble: 20,780 × 4
  date     state death positive
  <chr>    <chr> <dbl>    <dbl>
1 3/7/2021 AK      305    56886
2 3/7/2021 AL    10148   499819
3 3/7/2021 AR     5319   324818
4 3/7/2021 AS        0        0
5 3/7/2021 AZ    16328   826454
# ℹ 20,775 more rows


# <a name="Q9">Question 9</a>

---

Use the `separate()` function to create three new variables named `month`, `day` and `year` by separating the `date` column of `covid`.

- Assign the resulting data frame to a new data frame `covid_date`.
- Print the first 5 rows of `covid_date` to check your work.




In [43]:
# separate date into month, day, and year variables
# assign the result to covid_date
# print the first 5 rows to the screen

# Use separate() to separate the values from one column into multiple columns, where col is the column I'm taking the values from,
# into are the new column names, and sep is what is used to separate the values in the date column, and store it in covid_date
covid_date <- covid |> separate(col = date, into = c("month", "day", "year"), sep = "/")

# Print out the first 5 rows of the new data frame to check the result
print(covid_date, n = 5)

# A tibble: 20,780 × 6
  month day   year  state death positive
  <chr> <chr> <chr> <chr> <dbl>    <dbl>
1 3     7     2021  AK      305    56886
2 3     7     2021  AL    10148   499819
3 3     7     2021  AR     5319   324818
4 3     7     2021  AS        0        0
5 3     7     2021  AZ    16328   826454
# ℹ 20,775 more rows


# <a name="Q10">Question 10</a>

---


Use commands we have discussed this semester from the `base`, `dplyr` and/or `tidyr` packages to identify the range of dates included in the `covid_date` data frame.

Do not assign the output of your commands to any object, but be sure it is clear from the output displayed on the screen how you determined your answer.

- If you string together multiple operations, perform them in a single code cell using the pipe, `|>`.
- There should not be any command that you enter that prints more than 5 rows of output to the screen at any one time.
- The less output you need printed to the screen to answer this question, the better!
- Points will be deducted for displaying too much information on the screen to answer this question.
- Points will be deducted for using methods and/or commands we have not discussed this semester.

<mark>Clearly type your final answers in the text cell below.</mark>

<br>

## <a name="Sol10">Solution to Question 10</a>

---

- The start date of data in `covid_date` is 13th January, 2020.

- The end date of data in `covid_date` is 7th March, 2021.

<br>  

Run command(s) in code cell(s) below to generate output that supports your answers above. Be sure you do not display more than 5 rows of information to the screen as a result of any of your commands.

<br>  


In [44]:
# use code cell(s) to support your answers

# To find the start date, I used piping to string sequences of actions of finding the minimum value from year, month, and day, and print out only 1 row
covid_date |> slice_min(year) |> slice_min(month) |> slice_min(day) |> print(n = 1)

# A tibble: 1 × 6
  month day   year  state death positive
  <chr> <chr> <chr> <chr> <dbl>    <dbl>
1 1     13    2020  WA       NA       NA


In [45]:
# use code cell(s) to support your answers

# To find the end date, I used piping to string sequences of actions of finding the maximum value from year, month, and day, and print out only 1 row
covid_date |> slice_max(year) |> slice_max(month) |> slice_max(day) |> print(n = 1)

# A tibble: 56 × 6
  month day   year  state death positive
  <chr> <chr> <chr> <chr> <dbl>    <dbl>
1 3     7     2021  AK      305    56886
# ℹ 55 more rows


# <a name="Q11">Question 11</a>

---


Use the `filter()` function to subset the rows of `covid` for which the `date` is `3/7/2021`.  

- Now you are using the `covid` data frame from [Question 8](#Q8) with a single column for `date`.
  - Do NOT use the `covid_date` data frame from the previous question.
- Assign the resulting data frame back to the the name `covid`.
- Print the first 5 rows of `covid`.


In [46]:
# filter rows of covid with date 3/7/2021
# assign the result back to data frame named covid
# print the first 5 rows to the screen

# Use filter() to find the date that are equal to 3/7/2021, and store it in covid
covid <- covid |> filter(date == "3/7/2021")

# Print out the first 5 rows of the new data frame to check the result
print(covid, n = 5)

# A tibble: 56 × 4
  date     state death positive
  <chr>    <chr> <dbl>    <dbl>
1 3/7/2021 AK      305    56886
2 3/7/2021 AL    10148   499819
3 3/7/2021 AR     5319   324818
4 3/7/2021 AS        0        0
5 3/7/2021 AZ    16328   826454
# ℹ 51 more rows


# <a name="Q12">Question 12</a>

---



Use the `rename()` function to rename the `state` column of the `covid` data frame from the previous question to `state_abb`, the `death` column to `deaths`, and the `positive` column to `cases`.  

- Assign the resulting data frame back to the the name `covid`.
- Print the first 5 rows of `covid`.



In [47]:
# rename columns
# assign the result to covid
# print the first 5 rows to the screen

# Use rename(new_name = old_name) to rename each column's name, and store it back in covid
covid <- covid |> rename(state_abb = state, deaths = death, cases = positive)

# Print out the first 5 rows of the new data frame to check the result
print(covid, n = 5)

# A tibble: 56 × 4
  date     state_abb deaths  cases
  <chr>    <chr>      <dbl>  <dbl>
1 3/7/2021 AK           305  56886
2 3/7/2021 AL         10148 499819
3 3/7/2021 AR          5319 324818
4 3/7/2021 AS             0      0
5 3/7/2021 AZ         16328 826454
# ℹ 51 more rows


# <a name="Q13">Question 13</a>

---

We really need a `state` column in `covid` so that we can merge it with `state_demo`. But we only have a `state_abb` column.  The `state.name` and `state.abb` objects are available through `base` R for the 50 U.S. states.

- Run the first code cell below to create a data frame containing the state names and abbreviations of the 50 U.S. states and print the first 6 rows to the screen.

- Then answer the questions below.

In [48]:
# Ran this code to check out what state.name is (I was curious)
state.name

[1] "Alabama"        "Alaska"         "Arizona"        "Arkansas"      
 [5] "California"     "Colorado"       "Connecticut"    "Delaware"      
 [9] "Florida"        "Georgia"        "Hawaii"         "Idaho"         
[13] "Illinois"       "Indiana"        "Iowa"           "Kansas"        
[17] "Kentucky"       "Louisiana"      "Maine"          "Maryland"      
[21] "Massachusetts"  "Michigan"       "Minnesota"      "Mississippi"   
[25] "Missouri"       "Montana"        "Nebraska"       "Nevada"        
[29] "New Hampshire"  "New Jersey"     "New Mexico"     "New York"      
[33] "North Carolina" "North Dakota"   "Ohio"           "Oklahoma"      
[37] "Oregon"         "Pennsylvania"   "Rhode Island"   "South Carolina"
[41] "South Dakota"   "Tennessee"      "Texas"          "Utah"          
[45] "Vermont"        "Virginia"       "Washington"     "West Virginia" 
[49] "Wisconsin"      "Wyoming"

In [50]:
# creates data frame of state names and abbreviations

# This code was given by the professor on the Canvas announcement
state_labels <- tibble(state = state.name, state_abb = state.abb)

print(state_labels, n = 6)

# A tibble: 50 × 2
  state      state_abb
  <chr>      <chr>    
1 Alabama    AL       
2 Alaska     AK       
3 Arizona    AZ       
4 Arkansas   AR       
5 California CA       
6 Colorado   CO       
# ℹ 44 more rows


## <a name="Q13a">Question 13a</a>

---

How many rows of data are included in the `state_labels` data frame?

<br>  





### <a name="Sol13a">Solution to Question 13a</a>

---

Answer the question and explain how you used output from previous code cells above to answer the question.

There are 50 rows in this data frame.

You can see on the top it says tibble, and 50 x 2, and 50 is the number of rows, and 2 is the number of columns. Therefore, there are 50 rows in state_labels.

<br>  
<br>  


## <a name="Q13b">Question 13b</a>

---


Do the data frames `covid` and `state_labels` have all matching values in their `state_abb` columns?

- If yes, explain how you can determine both data frames have exactly the same values in the `state_abb` column.
- If not, explain which values (if any) are:
 - In the `state_abb` column of `covid` but not in the `state_abb` column of `state_labels`.
 - In the `state_abb` column of `state_labels` but not in the `state_abb` column of `covid`.




### <a name="Sol13b">Solution to Question 13b</a>

---
No, they do not have the same matching values.

In the covid data frame, there are 56 rows, which means the state_abb column in covid has 56 values. On the other hand, the state.abb column of state_labels only have 50 values, or 50 rows.



<br>  
<br>  


# <a name="Q14">Question 14</a>

---

Use either the `left_join()`, `right_join()`, `inner_join()`, or `full_join()` function to merge `state_labels` data frame into the `covid` data frame in the following way:

- You want to keep all rows and columns from the `covid` data frame.
- You want merge in the additional `state` column from the `state_labels` data frame.
- There will be some `NA` values filled in the column that is merged from the `state_labels` data frame.
- <font color="dodgerblue">**Assign the new data frame back to on object named `covid`**</font>.
- Print the first 9 rows of the resulting `covid` data frame to the screen.


The first 5 rows of the end result should look like the table below.

| date | state_abb | deaths | cases | state |
|------|-----------|--------|-------|-------|
| 3/7/2021 | AK | 305 | 56886 | Alaska |
| 3/7/2021 | AL | 10148 | 499819 | Alabama |
| 3/7/2021 | AR | 5319 | 324818 | Arkansas |
| 3/7/2021 | AS | 0 | 0 | NA |
| 3/7/2021 | AZ | 16328 | 826454 | Arizona |

You will print 9 rows of your resulting data frame to the screen which should end at Washington D.C.

<br>  

In [53]:
# use a *_join function to merge data frames
# assign result back to data frame named covid
# pring first 9 rows of resulting data frame

# Use left_join() to join the columns by having the "left" which is covid, as a main part, and add the information from state_labels to covid, and store it in covid
covid <- left_join(covid, state_labels)

# Print out the first 9 rows of the new data frame to check the result
print(covid, n = 9)

Joining with `by = join_by(state_abb, state)`


# A tibble: 56 × 5
  date     state_abb deaths   cases state      
  <chr>    <chr>      <dbl>   <dbl> <chr>      
1 3/7/2021 AK           305   56886 Alaska     
2 3/7/2021 AL         10148  499819 Alabama    
3 3/7/2021 AR          5319  324818 Arkansas   
4 3/7/2021 AS             0       0 NA         
5 3/7/2021 AZ         16328  826454 Arizona    
6 3/7/2021 CA         54124 3501394 California 
7 3/7/2021 CO          5989  436602 Colorado   
8 3/7/2021 CT          7704  285330 Connecticut
9 3/7/2021 DC          1030   41419 NA         
# ℹ 47 more rows


# <a name="Q15">Question 15</a>

---

The `state_demo` data frame has information for all 50 U.S. states and the District of Columbia. The `covid` data frame has information for all of these and more (including US territories). However, the state name for Washington, D.C. is not included in the current `covid` data frame (it is a missing `NA` value).  Run the code cell below to change the `state` value of Washington, D.C. to `District of Columbia` in the `covid` data frame.

- There is nothing for you to edit in the code cell below! Just run it.
- Assign the resulting data frame back to the the name `covid`.
- Print the first 9 rows of `covid`.



In [54]:
# nothing to edit in this code cell!
# run the code cell as is
covid <- covid |>
           mutate(state = replace(state, state_abb == "DC", "District of Columbia"))

# Print out the first 9 rows of the new data frame to check the result
print(covid, n = 9)

# A tibble: 56 × 5
  date     state_abb deaths   cases state               
  <chr>    <chr>      <dbl>   <dbl> <chr>               
1 3/7/2021 AK           305   56886 Alaska              
2 3/7/2021 AL         10148  499819 Alabama             
3 3/7/2021 AR          5319  324818 Arkansas            
4 3/7/2021 AS             0       0 NA                  
5 3/7/2021 AZ         16328  826454 Arizona             
6 3/7/2021 CA         54124 3501394 California          
7 3/7/2021 CO          5989  436602 Colorado            
8 3/7/2021 CT          7704  285330 Connecticut         
9 3/7/2021 DC          1030   41419 District of Columbia
# ℹ 47 more rows


# <a name="Q16">Question 16</a>

---

Use either the `left_join()`, `right_join()`, `inner_join()`, or `full_join()` function to merge `covid` data frame into the `state_demo` data frame in the following way:

- You want to keep all rows and columns from the `covid` data frame.
- You want merge in the additional `population`, `median_income`, `percent_hs_estimate`, and `percent_bs_estimate` columns from the `state_demo` data frame.
- There will be some `NA` values filled in the columns that are merged from the `state_demo` data frame since `state_demo` does not have any rows corresponding to US territories such as American Samoa (`AS`).
- <font color="dodgerblue">**Assign the new data frame back to on object named `covid`**</font>.
- Print the first 9 rows of the resulting `covid` data frame to the screen to check your work.


The first 5 rows of the end result should look like the table below. *Note the data in the last two columns, `percent_hs_estimate` and `percent_bs_estimate`, has been truncated in the table below.*

| date | state_abb | deaths | cases | state | population | median_income |
|------|-----------|--------|-------|-------|------------|--------------|
| 3/7/2021 | AK | 305 | 56886 | Alaska |  737068 | 35957 |
| 3/7/2021 | AL | 10148 | 499819 | Alabama | 4876250 | 26596 |
| 3/7/2021 | AR | 5319 | 324818 | Arkansas |  2999370 | 26137 |
| 3/7/2021 | AS | 0 | 0 | NA | NA | NA |
| 3/7/2021 | AZ | 16328 | 826454 | Arizona |  7050299 | 30552 |

You will print 9 rows of your resulting data frame to the screen which should end at Washington D.C.


<br>  

In [55]:
# use a *_join function to merge data frames
# assign result back to data frame named covid
# pring first 9 rows of resulting data frame

# Use left_join() to join the columns by having the "left" which is covid, as a main part, and add the information from state_demo to covid, and store it in covid
covid <- left_join(covid, state_demo)

# Print out the first 9 rows of the new data frame to check the result
print(covid, n = 9)

Joining with `by = join_by(state)`


# A tibble: 56 × 9
  date     state_abb deaths   cases state               population median_income
  <chr>    <chr>      <dbl>   <dbl> <chr>                    <dbl>         <dbl>
1 3/7/2021 AK           305   56886 Alaska                  737068         35957
2 3/7/2021 AL         10148  499819 Alabama                4876250         26596
3 3/7/2021 AR          5319  324818 Arkansas               2999370         26137
4 3/7/2021 AS             0       0 NA                          NA            NA
5 3/7/2021 AZ         16328  826454 Arizona                7050299         30552
6 3/7/2021 CA         54124 3501394 California            39283497         32232
7 3/7/2021 CO          5989  436602 Colorado               5610349         36644
8 3/7/2021 CT          7704  285330 Connecticut            3575074         38809
9 3/7/2021 DC          1030   41419 District of Columb…     692683         50134
# ℹ 47 more rows
# ℹ 2 more variables: percent_hs_estimate <dbl>, percent_bs_estimate <dbl

# <a name="Q17">Question 17</a>

---

When manipulating data into a usable form, it can be convenient to save the final object (or even intermediate objects) to easily load for future use. This is the purpose of the `write_*()` and `read_*()` functions in the `readr` package if we are in the `tidyverse` or the `read.*` and `write.*` functions in `base` R as another alternative. Since we are working in the `tidyverse`, we use functions in the `readr` package.


- We used the `read_csv()` function earlier in this assignment to load external csv files into an R session.
- We can similarly use the `write_csv()` function to export a data frame to a csv file that we save externally.


Run the code cell below to export your final, tidy `covid` data frame to a csv file name `covid-hw6.csv`.




In [56]:
# run the code cell
# there is nothing to edit
# no output will be printed to the screen
write_csv(covid, "covid_hw6.csv")

# <a name="Q18">Question 18</a>

---

Where did the file `covid_hw6.csv` that we created go? Where is it saved? Right now, it is stored in a rather hidden location in your Google Drive. Here is how to find the file and download a copy locally on to your computer.

<br>

<img src="https://drive.google.com/uc?export=view&id=1sAWqe7s_qmhrnHGF7ABdQN27ir4H9KoA"  alt = "Download csv file" width="300">


- Select the files menu tab from the options on the left side bar (the image of the file boxed in red in the figure above).
- You should now see the file `covid_hw6.csv` listed in your available files.
- Click on the three vertical dots on the right of the file name (see red arrow in figure above).
- Choose `Download` from the available menu options.
- Choose where you want to download the file onto your computer.

Follow the process above to download a local copy of the file `covid_hw6.csv` onto your computer.

**Upload your final csv file named `covid_hw6.csv` along with your `*.ipynb` file in Canvas when you submit your work for [Homework 6 in Canvas](https://ucdenver.instructure.com/courses/558344/assignments/1920900). Thus, we you submit you work for Homework 6, you should submit two files:**

1. A `*.ipynb` which is a copy of this file with your solutions and all code run with the output viewable.
2. The csv file `covid_hw6.csv` you generated from this assignment.


<br>  



# <a name="save2">Be Sure You Have Saved Your Work!</a>

---

This notebook is a shared Colab notebook available for anyone to view. However, since everyone is sharing this notebook, you do not have permission to save changes to this shared Colab notebook. In order to save your work:

1. You will need to set up a free Google Drive account. If you already have a Google Drive account, you are ready to go!
2. Click the `Copy to Drive` button to the right of the `+ Code` and `+ Text` buttons on top of the notebook.
3. Select from the menu `File/Save a Copy in Drive`.
  - By default, the notebook will be saved in a folder named **Colab Notebooks** in your Drive.
  - Feel free to rename and store the notebook wherever you like.
